### PR TITLE
Reorganize profiles into their own files

### PR DESCRIPTION
**Pull request by TheSpydog**
Couldn't get to sleep last night, so I did what any normal person would do -- clean up MojoShader! I ripped out all the profiles and put them in their own files in the `profiles/` subdirectory. IMO this has been long overdue, but after having done it, I can totally see why nobody attempted it before. :P

A few things:
1. This works just fine in all of my tests, but it'd be good to make doubly sure I didn't break anything in the transition.
2. It might be worth a pass over the new header files to make sure the formatting looks okay. Beauty wasn't a very high priority when working on this initial commit...
3. We've got to figure out what to do about that `mojoshader_internal.h` error that I commented out. It's being fussy with the new files (probably due to some preprocessor ordering rule I'm not familiar with), but hopefully it won't be a difficult fix.
4. Since we're already overhauling the project structure, what are your thoughts on killing off that ARB1 target? TBH I don't even know what it is -- the top google search for "arb1 shader" was MojoShader's website. But I do know that removing it would simplify some things and a fix a [small weird bit](https://github.com/FNA-XNA/MojoShader/compare/fna...TheSpydog:fna#diff-f6bd4f8f8bd83197cca2564b822858d9R320) in the current implementation.

### CODE REVIEW

**Comment by flibitijibibo** [comment]
1. I’ll most likely start with a side-by-side comparison along with a binary size comparison. I’m hoping it’ll be exactly the same without debug symbols...
2. Any spots that are all new should be noted, just so we know what was actually added compared to the original.
3. Probably just need to define that value before including stuff in the profile c files.
4. I thought about this a lot, but only because the main file was so huge. ARB1 refers to ARB_vertex_program and ARB_fragment program, along with some NV shader extensions. Most (if not all) of this predates GLSL! This is how you can tell MojoShader was written in 2008... back then you needed these for good hardware support, but that’s not really true anymore. But I know Ryan likes his old hardware support, so I dunno if I want to mess with it. At least now it’s not as annoying to maintain.

**Comment by TheSpydog** [comment]
1. From my testing, the binary size (even when stripped) is quite a bit bigger than the previous version. I'm wondering if it's due to moving a lot of `static inline` functions out of mojoshader.c and into the mojoshader_profile_common.h where they're getting copied and pasted around several files.
2. The profile header files are the only major addition, but they were copied directly from their corresponding source files (with small changes to macros and such). Beyond that, I don't think there was anything significant added...
3. Yup, that was it.
4. Makes sense!

**Comment by flibitijibibo** [comment]
Very well, I’ll try to review the refactor tomorrow.

We’ll have to figure out the best path for dealing with the static functions. May have to macro, may have to throw them in common.c. That may have to be a separate commit though, since this changes enough as it is...

File: profiles/mojoshader_profile.h [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+    emit_uniform uniform_emitter;
+    emit_sampler sampler_emitter;
+    emit_attribute attribute_emitter;
+    emit_finalize finalize_emitter;
+    varname_function get_varname;
+    const_array_varname_function get_const_array_varname;
+} Profile;
+
+
+// Common utilities...
```
Comment (flibitijibibo) [inline profiles/mojoshader_profile.h:292]: Marking this as a change of mojoshader.c:301...

File: profiles/mojoshader_profile.h [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+        strcpy(retval, str);
+    return retval;
+} // StrDup
+
+static inline void Free(Context *ctx, void *ptr)
+{
+    ctx->free(ptr, ctx->malloc_data);
+} // Free
+
+int set_output(Context *ctx, Buffer **section);
```
Comment (flibitijibibo) [inline profiles/mojoshader_profile.h:301]: Marking this as a change of mojoshader.c:342...

File: profiles/mojoshader_profile.h [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+    return (ctx->shader_type == MOJOSHADER_TYPE_VERTEX);
+} // shader_is_vertex
+
+
+static inline int isfail(const Context *ctx)
+{
+    return ctx->isfail;
+} // isfail
+
+void failf(Context *ctx, const char *fmt, ...);
```
Comment (flibitijibibo) [inline profiles/mojoshader_profile.h:315]: Marking this as a change of mojoshader.c:420...

File: profiles/mojoshader_profile.h [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+{
+    assert(ctx->output != NULL);
+    if (!isfail(ctx))
+        buffer_append(ctx->output, ctx->endline, ctx->endline_len);
+} // output_blank_line
+
+void floatstr(Context *ctx, char *buf, size_t bufsize, float f,
+                     int leavedecimal);
+
+RegisterList *reglist_insert(Context *ctx, RegisterList *prev,
```
Comment (flibitijibibo) [inline profiles/mojoshader_profile.h:324]: Seeing lots of changes here compared to mojoshader.c:516...

File: profiles/mojoshader_profile.h [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+static inline void set_dstarg_writemask(DestArgInfo *dst, const int mask)
+{
+    dst->writemask = mask;
+    dst->writemask0 = ((mask >> 0) & 1);
+    dst->writemask1 = ((mask >> 1) & 1);
+    dst->writemask2 = ((mask >> 2) & 1);
+    dst->writemask3 = ((mask >> 3) & 1);
+} // set_dstarg_writemask
+
+int allocate_scratch_register(Context *ctx);
```
Comment (flibitijibibo) [inline profiles/mojoshader_profile.h]: Marking as a change of mojoshader.c:743...

File: profiles/mojoshader_profile_d3d.c [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+ *  This file written by Ryan C. Gordon.
+ */
+
+#define __MOJOSHADER_INTERNAL__ 1
+#include "mojoshader_profile.h"
+
+#if SUPPORT_PROFILE_D3D
+
+const char *make_D3D_srcarg_string_in_buf(Context *ctx,
+                                                 const SourceArgInfo *arg,
```
Comment (flibitijibibo) [inline profiles/mojoshader_profile_d3d.c]: Going to be lots of these extra spaces since the functions are no longer static.

File: mojoshader.c [hunk 7]
```diff
@@ -8621,6 +324,7 @@ static const struct { const char *from; const char *to; } profileMap[] =
      PROFILE_EMITTER_METAL(op) \
 }
 
+
```
Comment (flibitijibibo) [inline mojoshader.c]: Extra space!

File: mojoshader.c [hunk 8]
```diff
@@ -16,8554 +16,257 @@
...
-
-    if (!arg->relative)
-    {
-        regtype_str = get_GLSL_varname_in_buf(ctx, arg->regtype, arg->regnum,
-                                              (char *) alloca(64), 64);
-    } // if
-
-    const char *rel_lbracket = "";
-    char rel_offset[32] = { '\0' };
-    const char *rel_rbracket
```
Comment (flibitijibibo) [inline mojoshader.c:114]: What we can do here is move this block above the `!AT_LEAST_ONE_PROFILE` check, then change this...

```
#if !SUPPORT_PROFILE_GLSL
#define PROFILE_EMITTER_GLSL(op)
#else
#undef AT_LEAST_ONE_PROFILE
#define AT_LEAST_ONE_PROFILE 1
#define PROFILE_EMITTER_GLSL(op) emit_GLSL_##op,
#endif
```

... into this:

```
#if SUPPORT_PROFILE_GLSL
#undef AT_LEAST_ONE_PROFILE
#define AT_LEAST_ONE_PROFILE 1
#define PROFILE_EMITTER_GLSL(op) emit_GLSL_op,
PREDECLARE_PROFILE(GLSL)
#endif // SUPPORT_PROFILE_GLSL
```

Mostly trying to compress the steps needed to add a new profile. Current steps by my count:

1. Add new C file to profiles/, implement emitter
2. Add file to CMakeLists.txt
3. Add PROFILE_EMITTER_NEWPROFILE block to mojoshader.c
4. Add PREDECLARE_PROFILE block to mojoshader.c
5. Add new profile to `profiles` in mojoshader.c
6. Add to PROFILE_EMITTERS macro in mojoshader.c
7. Add to MOJOSHADER_maxShaderModel in mojoshader.c

Pretty much all of these aren't our fault, but I bet we can compress 3 and 4 into one step. If we ever remove the ARB1 group then we'll be able to get rid of step 7 as well.

**Review by flibitijibibo — COMMENTED** [review]
This was my first review pass. Overall it seems fine aside from style things, but I'm not crazy enough to expect perfect style from a 17k-line patch.

The main thing we need to document very very carefully in our final commit message is where everything moved. I tried marking spots that were clear transfers but then things fell apart a little since stuff got moved into profile_common, or were actually only used in certain spots, and some functions stayed as static inline while others did not, etc.

What might be the best next step is starting the process of replacing all the `static inline` functions with whatever is more appropriate now that it's all split up. Some of it could be macro-fied (i.e. all the one-liners) but that would mean editing even _more_ code, so let's do that another time. What we _can_ do now is try to group the related functions together and try to give them the same attributes. For example, if we have this...

```
static inline void Blorp1()
{
    lots;
    of;
    if (crap) {
       oh;
       good;
       grief;
    } // if
} // Blorp1

static inline int Blorp2(int zoop)
{
    return zoop == 69;
} // Blorp2

static inline int Blorp3(const char *why)
{
    return strcmp(why, "because");
} // Blorp3
```

... even though Blorp2/3 are trivial to inline, we have to be the bad guy and put all the Blorps in common.c together, instead of mixing/matching. Thankfully the only place where this got really messy was the register functions, everything else seemed clear.

The final version of this one patch will probably make the binary size a lot bigger, but fixing that can be a separate commit.

File: mojoshader.c (hunk 8 above)
Comment (TheSpydog) [reply mojoshader.c:114]: Which block are you referring to? I don't think the line linked with this comment is related to your suggestion.

Comment (flibitijibibo) [reply mojoshader.c:114]: Oh crap, it got moved. Here: https://github.com/TheSpydog/MojoShader/blob/fna/mojoshader.c#L161

**Comment by TheSpydog** [comment]
All right, all the formerly `static inline` functions from mojoshader_profile.h are now non-static, non-inline functions housed in mojoshader_profile_common.c. I've additionally tried to organize things a bit better to match the original declaration order of mojoshader.c, although there are still some notable changes and/or omissions:

Several functions are only ever used in mojoshader.c, so they're still defined as static functions in that file:
- `free_reglist()`
- `reglist_exists()`
- `register_was_written()`
- `get_defined_register()`
- `add_attribute_register()`
- `add_sampler()`
- `adjust_token_position()`

A couple functions turned out to only be used by the ARB1 profile, so I moved them into mojoshader_profile_arb1.c:
- `allocate_scratch_register()`
- `allocate_branch_label()`

And finally, `get_used_register()` was literally never used anywhere, so I just removed it.

Also, I merged the `PREDECLARE_PROFILE` macro calls into the existing `#if SUPPORT_PROFILE_*` stuff, per your comment.

**Comment by TheSpydog** [comment]
All right, with the help of [Bloaty McBloatface](https://github.com/google/bloaty) and `#pragma GCC visibility`, I've managed to reduce the size difference between this version and the old version to a mere 56 bytes (when stripped and compiled with `-DCOMPILER_SUPPORT=OFF`).

Other notable change: `usagestrs[]` has been moved from `mojoshader_profile.h` to `mojoshader_profile_d3d.c` since only the D3D profile made use of this array.

**Comment by TheSpydog** [comment]
What's left to do here before this can be merged? Are we waiting on the SPIR-V emitter so we can refactor that into the new format too?

**Comment by flibitijibibo** [comment]
I'll do one last review of this today and if everything looks good we can make a Mercurial patch to push to upstream. EDIT: Have to push this to tomorrow, sorry!

File: profiles/mojoshader_profile.h [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+int writemask_y(const int writemask);
+int replicate_swizzle(const int swizzle);
+int no_swizzle(const int swizzle);
+int vecsize_from_writemask(const int m);
+void set_dstarg_writemask(DestArgInfo *dst, const int mask);
+
+int isscalar(Context *ctx, const MOJOSHADER_shaderType shader_type,
+             const RegisterType rtype, const int rnum);
+
+static const char swizzle_channels[] = { 'x', 'y', 'z', 'w' };
```
Comment (flibitijibibo) [inline profiles/mojoshader_profile.h:350]: May be a possible cause of increased size, but just barely...

Comment (TheSpydog) [reply profiles/mojoshader_profile.h:350]: Yeah, it's annoying. I've tried moving it around and making it non-static, but no matter what it doesn't seem to affect the stripped binary size. :/

**Comment by flibitijibibo** [comment]
Passed this through all of my compilers and they seem to be happy with everything. The very last thing we get to do also happens to be the most tedious: We'll need to fix parameter lists for functions that used to be `static` but aren't anymore. One example is `isscalar`, which looked like this...

https://github.com/FNA-XNA/MojoShader/pull/5/files#diff-d496920d66adab7260e5d2670f334605L766

... but now looks like this:

https://github.com/FNA-XNA/MojoShader/pull/5/files#diff-f6bd4f8f8bd83197cca2564b822858d9R357

We just have to line everything up again. Once that's done we should be good to go on the Hg version of this patch.

File: profiles/mojoshader_profile.h (hunk 9 above)
Comment (flibitijibibo) [reply profiles/mojoshader_profile.h:350]: That's hopefully the optimizer being smart. Oh well, we can just leave this...

**Comment by TheSpydog** [comment]
All right, I think I got 'em all. There could be a few that snuck by me, but we'll fix them as we see them.

Here's the hg patch (uploaded as a .txt file since GitHub doesn't like .patch files):
[separateprofiles.txt](https://github.com/FNA-XNA/MojoShader/files/3109057/separateprofiles.txt)

**Comment by flibitijibibo** [comment]
Merged into upstream:

https://hg.icculus.org/icculus/mojoshader/rev/b8ece252a201
https://github.com/FNA-XNA/MojoShader/commit/cd04fa3f1296f9aa99008b2310b5eabc3558a0d5

**Pull request closed without merging**